capsules/extra/src/nmea.rs

```diff
@@ -0,0 +1,126 @@
+use kernel::{ErrorCode, ProcessId};
+
+/// Syscall driver number.
+use capsules_core::driver;
+pub const DRIVER_NUM: usize = driver::NUM::Nmea as usize;
+
+#[derive(Clone, Copy, PartialEq, Default)]
+enum Operation {
+    #[default]
+    None,
+    Read,
+}
+
+#[derive(Default)]
+pub struct App {
+    operation: Operation,
+}
+
+pub struct Nmea<'a> {
+    driver: &'a dyn hil::sensors::NmeaDriver<'a>,
+    apps: Grant<App, UpcallCount<1>, AllowRoCount<0>, AllowRwCount<1>>,
+    busy: Cell<bool>,
+    buffer: TakeCell<'static, [u8]>,
+}
+
+impl<'a> Nmea<'a> {
+    pub fn new(
+        driver: &'a dyn hil::sensors::NmeaDriver<'a>,
+        grant: Grant<App, UpcallCount<1>, AllowRoCount<0>, AllowRwCount<1>>,
+        buffer: &'static mut [u8],
+    ) -> Nmea<'a> {
+        Nmea {
+            driver,
+            apps: grant,
+            busy: Cell::new(false),
+            buffer: TakeCell::new(buffer),
+        }
+    }
+}
+
+impl hil::sensors::NmeaClient for Nmea<'_> {
+    fn callback(&self, buffer: &'static mut [u8], len: usize, status: Result<(), i2c::Error>) {
+        for cntr in self.apps.iter() {
+            cntr.enter(|app, kernel_data| {
+                if app.operation == Operation::Read {
+                    if status.is_err() {
+                        kernel_data
+                            .schedule_upcall(0, (into_statuscode(Err(ErrorCode::FAIL)), 0, 0))
+                            .ok();
+                    }
+
+                    let _ = kernel_data.get_readwrite_processbuffer(0).and_then(|dest| {
+                        dest.mut_enter(|dest| {
+                            let copy_len = dest.len().max(len);
+
+                            dest[0..copy_len].copy_from_slice(&buffer[0..copy_len]);
+                        })
+                    });
+
+                    self.busy.set(false);
+
+                    app.operation = Operation::None;
+                    kernel_data.schedule_upcall(0, (0, len, 0)).ok();
+                }
+            });
+        }
+
+        self.buffer.replace(buffer);
+    }
+}
+
+impl SyscallDriver for Nmea<'_> {
+    fn command(
+        &self,
+        command_num: usize,
+        _: usize,
+        _: usize,
+        processid: ProcessId,
+    ) -> CommandReturn {
+        match command_num {
+            // check whether the driver exists!!
+            0 => CommandReturn::success(),
+
+            // Read sentance
+            1 => self
+                .apps
+                .enter(processid, |app, _| {
+                    if !self.busy.get() {
+                        self.busy.set(true);
+                        app.operation = Operation::Read;
+
+                        match self.driver.read_sentance(self.buffer.take().unwrap()) {
+                            Ok(()) => CommandReturn::success(),
+                            Err((e, buffer)) => {
+                                self.buffer.replace(buffer);
+                                CommandReturn::failure(e)
```

Comment:
I think this needs to reset the `app.operation` and `self.busy` in the case of a failure